File: AssetManager/UserInterface/Forms/AdminTools/UserManagerForm.vb
﻿Public Class UserManagerForm
    Private ModuleIndex As New List(Of AccessGroupObject)
    Private CurrentUser As LocalUserInfoStruct
    Private UserDataQuery As String = "SELECT * FROM " & UsersCols.TableName
    Private SelectedRow As Integer

    Sub New(parentForm As ExtendedForm)
        InitializeComponent()
        Me.ParentForm = parentForm
        Show()
    End Sub

    Private Sub frmUserManager_Load(sender As Object, e As EventArgs) Handles Me.Load
        LoadUserData()
    End Sub

    Private Sub LoadUserData()
        ListUsers()
        ModuleIndex = BuildModuleIndex()
        LoadModuleBoxes()
        UpdateAccessLabel()
    End Sub

    Private Sub ListUsers()
        SendToGrid()
    End Sub

    Private Sub SendToGrid()
        UserGrid.DataSource = DBFactory.GetDatabase.DataTableFromQueryString(UserDataQuery)
        UserGrid.Columns(UsersCols.UID).ReadOnly = True
    End Sub

    Private Sub DisplayAccess(intAccessLevel As Integer)
        Dim clbItemStates(clbModules.Items.Count - 1) As CheckState
        For Each chkBox As CheckBox In clbModules.Items
            If SecurityTools.CanAccess(chkBox.Name, intAccessLevel) Then
                clbItemStates(clbModules.Items.IndexOf(chkBox)) = CheckState.Checked
            Else
                clbItemStates(clbModules.Items.IndexOf(chkBox)) = CheckState.Unchecked
            End If
        Next
        For i As Integer = 0 To clbItemStates.Count - 1
            clbModules.SetItemCheckState(i, clbItemStates(i))
        Next
        UpdateAccessLabel()
        AutoSizeCLBColumns(clbModules)
    End Sub

    Public Function BuildModuleIndex() As List(Of AccessGroupObject)
        Dim tmpList As New List(Of AccessGroupObject)
        Using ModuleTable As DataTable = DBFactory.GetDatabase.DataTableFromQueryString("SELECT * FROM " & SecurityCols.TableName & " ORDER BY " & SecurityCols.AccessLevel & "")
            For Each row As DataRow In ModuleTable.Rows
                tmpList.Add(New AccessGroupObject(row))
            Next
            Return tmpList
        End Using
    End Function

    Private Sub LoadModuleBoxes()
        Dim chkModuleBox As CheckBox
        clbModules.Items.Clear()
        For Each ModuleBox As AccessGroupObject In ModuleIndex
            chkModuleBox = New CheckBox
            With chkModuleBox
                .Text = ModuleBox.Description
                .Name = ModuleBox.AccessModule
            End With
            clbModules.DisplayMember = "Text"
            clbModules.Items.Add(chkModuleBox)
        Next
        AutoSizeCLBColumns(clbModules)
    End Sub

    Private Function CalcAccessLevel() As Integer
        Dim intAccessLevel As Integer = 0
        For Each chkBox As CheckBox In clbModules.Items
            If clbModules.GetItemCheckState(clbModules.Items.IndexOf(chkBox)) = CheckState.Checked Then
                intAccessLevel += SecurityTools.GetSecGroupValue(chkBox.Name)
            End If
        Next
        Return intAccessLevel
    End Function

    Private Sub UserGrid_CellClick(sender As Object, e As DataGridViewCellEventArgs) Handles UserGrid.CellClick
        If TypeOf UserGrid.Item(GridFunctions.GetColIndex(UserGrid, UsersCols.AccessLevel), UserGrid.CurrentRow.Index).Value Is Integer Then
            DisplayAccess(CInt(UserGrid.Item(GridFunctions.GetColIndex(UserGrid, UsersCols.AccessLevel), UserGrid.CurrentRow.Index).Value))
            GetUserInfo()
        Else
            DisplayAccess(0)
        End If
        SelectedRow = e.RowIndex
    End Sub

    Private Sub GetUserInfo()
        With CurrentUser
            .AccessLevel = CInt(GridFunctions.GetCurrentCellValue(UserGrid, UsersCols.AccessLevel))
            .UserName = GridFunctions.GetCurrentCellValue(UserGrid, UsersCols.UserName)
            .GUID = GridFunctions.GetCurrentCellValue(UserGrid, UsersCols.UID)
            .Fullname = GridFunctions.GetCurrentCellValue(UserGrid, UsersCols.FullName)
        End With
    End Sub

    Private Sub cmdUpdate_Click(sender As Object, e As EventArgs) Handles cmdUpdate.Click
        Try
            Dim blah = Message("Are you sure?  Committed changes cannot be undone.", vbYesNo + vbQuestion, "Commit Changes", Me)
            If blah = DialogResult.Yes Then
                UserGrid.EndEdit()
                AddGUIDs()
                DBFactory.GetDatabase.UpdateTable(UserDataQuery, DirectCast(UserGrid.DataSource, DataTable))
                ListUsers()
                SecurityTools.GetUserAccess()
            Else

            End If
        Catch ex As Exception
            ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod())
        End Try
    End Sub

    Private Sub AddGUIDs()
        For Each rows As DataGridViewRow In UserGrid.Rows
            If rows.Cells(GridFunctions.GetColIndex(UserGrid, UsersCols.UID)).EditedFormattedValue.ToString = "" Then
                Dim UserUID As String = Guid.NewGuid.ToString
                rows.Cells(GridFunctions.GetColIndex(UserGrid, UsersCols.UID)).Value = UserUID
            End If
        Next
    End Sub

    Private Sub AddAccessLevelToGrid()
        UserGrid.Rows(SelectedRow).Cells(GridFunctions.GetColIndex(UserGrid, UsersCols.AccessLevel)).Selected = True
        UserGrid.BeginEdit(False)
        UserGrid.Rows(SelectedRow).Cells(GridFunctions.GetColIndex(UserGrid, UsersCols.AccessLevel)).Value = CalcAccessLevel()
        UserGrid.EndEdit()
    End Sub

    Private Sub AutoSizeCLBColumns(CLB As CheckedListBox)
        CLB.ColumnWidth = CInt(CLB.Width / 2)
    End Sub

    Private Sub UpdateAccessLabel()
        lblAccessValue.Text = "Selected Access Level: " & CalcAccessLevel()
    End Sub

    Private Sub clbModules_MouseUp(sender As Object, e As MouseEventArgs) Handles clbModules.MouseUp
        UpdateAccessLabel()
        AddAccessLevelToGrid()
    End Sub

    Private Sub UserGrid_KeyDown(sender As Object, e As KeyEventArgs) Handles UserGrid.KeyDown
        If e.KeyCode = Keys.Delete Then
            UserGrid.Rows.RemoveAt(SelectedRow)
        End If
    End Sub

    Private Sub UserGrid_CellDoubleClick(sender As Object, e As DataGridViewCellEventArgs) Handles UserGrid.CellDoubleClick
        UserGrid.BeginEdit(False)
    End Sub

    Private Sub cmdRefresh_Click(sender As Object, e As EventArgs) Handles cmdRefresh.Click
        LoadUserData()
    End Sub

End Class

File: AssetManager/UserInterface/Forms/AdminTools/UserManagerForm.cs
using AssetManager.UserInterface.CustomControls;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Windows.Forms;

namespace AssetManager.UserInterface.Forms.AdminTools
{
    public partial class UserManagerForm : ExtendedForm
    {
        private List<AccessGroupObject> ModuleIndex = new List<AccessGroupObject>();
        private LocalUserInfoStruct CurrentUser;
        private string UserDataQuery = "SELECT * FROM " + UsersCols.TableName;

        private int SelectedRow;

        public UserManagerForm(ExtendedForm parentForm)
        {
            Load += frmUserManager_Load;
            InitializeComponent();
            this.ParentForm = parentForm;
            StyleFunctions.SetGridStyle(UserGrid, GridTheme);
            Show();
        }

        private void frmUserManager_Load(object sender, EventArgs e)
        {
            LoadUserData();
        }

        private void LoadUserData()
        {
            ListUsers();
            ModuleIndex = BuildModuleIndex();
            LoadModuleBoxes();
            UpdateAccessLabel();
        }

        private void ListUsers()
        {
            SendToGrid();
        }

        private void SendToGrid()
        {
            UserGrid.DataSource = DBFactory.GetDatabase().DataTableFromQueryString(UserDataQuery);
            UserGrid.Columns[UsersCols.UID].ReadOnly = true;
        }

        private void DisplayAccess(int intAccessLevel)
        {
            CheckState[] clbItemStates = new CheckState[clbModules.Items.Count];
            foreach (CheckBox chkBox in clbModules.Items)
            {
                if (SecurityTools.CanAccess(chkBox.Name, intAccessLevel))
                {
                    clbItemStates[clbModules.Items.IndexOf(chkBox)] = CheckState.Checked;
                }
                else
                {
                    clbItemStates[clbModules.Items.IndexOf(chkBox)] = CheckState.Unchecked;
                }
            }
            for (int i = 0; i <= clbItemStates.Count() - 1; i++)
            {
                clbModules.SetItemCheckState(i, clbItemStates[i]);
            }
            UpdateAccessLabel();
            AutoSizeCLBColumns(clbModules);
        }

        public List<AccessGroupObject> BuildModuleIndex()
        {
            List<AccessGroupObject> tmpList = new List<AccessGroupObject>();
            using (DataTable ModuleTable = DBFactory.GetDatabase().DataTableFromQueryString("SELECT * FROM " + SecurityCols.TableName + " ORDER BY " + SecurityCols.AccessLevel + ""))
            {
                foreach (DataRow row in ModuleTable.Rows)
                {
                    tmpList.Add(new AccessGroupObject(row));
                }
                return tmpList;
            }
        }

        private void LoadModuleBoxes()
        {
            CheckBox chkModuleBox = null;
            clbModules.Items.Clear();
            foreach (AccessGroupObject ModuleBox in ModuleIndex)
            {
                chkModuleBox = new CheckBox();
                var _with1 = chkModuleBox;
                _with1.Text = ModuleBox.Description;
                _with1.Name = ModuleBox.AccessModule;
                clbModules.DisplayMember = "Text";
                clbModules.Items.Add(chkModuleBox);
            }
            AutoSizeCLBColumns(clbModules);
        }

        private int CalcAccessLevel()
        {
            int intAccessLevel = 0;
            foreach (CheckBox chkBox in clbModules.Items)
            {
                if (clbModules.GetItemCheckState(clbModules.Items.IndexOf(chkBox)) == CheckState.Checked)
                {
                    intAccessLevel += SecurityTools.GetSecGroupValue(chkBox.Name);
                }
            }
            return intAccessLevel;
        }

        private void UserGrid_CellClick(object sender, DataGridViewCellEventArgs e)
        {
            if (UserGrid[GridFunctions.GetColIndex(UserGrid, UsersCols.AccessLevel), UserGrid.CurrentRow.Index].Value is int)
            {
                DisplayAccess(Convert.ToInt32(UserGrid[GridFunctions.GetColIndex(UserGrid, UsersCols.AccessLevel), UserGrid.CurrentRow.Index].Value));
                GetUserInfo();
            }
            else
            {
                DisplayAccess(0);
            }
            SelectedRow = e.RowIndex;
        }

        private void GetUserInfo()
        {
            var _with2 = CurrentUser;
            _with2.AccessLevel = Convert.ToInt32(GridFunctions.GetCurrentCellValue(UserGrid, UsersCols.AccessLevel));
            _with2.UserName = GridFunctions.GetCurrentCellValue(UserGrid, UsersCols.UserName);
            _with2.GUID = GridFunctions.GetCurrentCellValue(UserGrid, UsersCols.UID);
            _with2.Fullname = GridFunctions.GetCurrentCellValue(UserGrid, UsersCols.FullName);
        }

        private void cmdUpdate_Click(object sender, EventArgs e)
        {
            try
            {
                var blah = OtherFunctions.Message("Are you sure?  Committed changes cannot be undone.", (int)MessageBoxButtons.YesNo + (int)MessageBoxIcon.Question, "Commit Changes", this);
                if (blah == DialogResult.Yes)
                {
                    UserGrid.EndEdit();
                    AddGUIDs();
                    DBFactory.GetDatabase().UpdateTable(UserDataQuery, (DataTable)UserGrid.DataSource);
                    ListUsers();
                    SecurityTools.GetUserAccess();
                }
                else
                {
                }
            }
            catch (Exception ex)
            {
                ErrorHandling.ErrHandle(ex, System.Reflection.MethodInfo.GetCurrentMethod());
            }
        }

        private void AddGUIDs()
        {
            foreach (DataGridViewRow rows in UserGrid.Rows)
            {
                if (string.IsNullOrEmpty(rows.Cells[GridFunctions.GetColIndex(UserGrid, UsersCols.UID)].EditedFormattedValue.ToString()))
                {
                    string UserUID = Guid.NewGuid().ToString();
                    rows.Cells[GridFunctions.GetColIndex(UserGrid, UsersCols.UID)].Value = UserUID;
                }
            }
        }

        private void AddAccessLevelToGrid()
        {
            UserGrid.Rows[SelectedRow].Cells[GridFunctions.GetColIndex(UserGrid, UsersCols.AccessLevel)].Selected = true;
            UserGrid.BeginEdit(false);
            UserGrid.Rows[SelectedRow].Cells[GridFunctions.GetColIndex(UserGrid, UsersCols.AccessLevel)].Value = CalcAccessLevel();
            UserGrid.EndEdit();
        }

        private void AutoSizeCLBColumns(CheckedListBox CLB)
        {
            CLB.ColumnWidth = Convert.ToInt32(CLB.Width / 2);
        }

        private void UpdateAccessLabel()
        {
            lblAccessValue.Text = "Selected Access Level: " + CalcAccessLevel();
        }

        private void clbModules_MouseUp(object sender, MouseEventArgs e)
        {
            UpdateAccessLabel();
            AddAccessLevelToGrid();
        }

        private void UserGrid_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Delete)
            {
                UserGrid.Rows.RemoveAt(SelectedRow);
            }
        }

        private void UserGrid_CellDoubleClick(object sender, DataGridViewCellEventArgs e)
        {
            UserGrid.BeginEdit(false);
        }

        private void cmdRefresh_Click(object sender, EventArgs e)
        {
            LoadUserData();
        }
    }
}